relay-server/src/actors/store.rs

```diff
@@ -199,51 +199,101 @@ impl StoreService {
-                    event_id: event_id.ok_or(StoreError::NoEventId)?,
-                    project_id: scoping.project_id,
-                    attachment,
-                });
+        let remote_addr = envelope.meta().client_addr().map(|addr| addr.to_string());
 
-                self.produce(topic, scoping.organization_id, attachment_message)?;
-                metric!(
-                    counter(RelayCounters::ProcessingMessageProduced) += 1,
-                    event_type = "attachment"
-                );
+        let kafka_messages = Self::extract_kafka_messages(
+            event_item,
+            event_id.ok_or(StoreError::NoEventId)?,
+            scoping,
+            start_time,
+            remote_addr,
+            attachments,
+        );
+
+        for message in kafka_messages {
+            if let Ok(()) = self.produce(topic, scoping.organization_id, &message) {
+                // If message was succesfully sent, make the right log message.
+                match message {
+                    KafkaMessage::Attachment(_) => {
+                        metric!(
+                            counter(RelayCounters::ProcessingMessageProduced) += 1,
+                            event_type = "attachment"
+                        );
+                    }
+                    _ => {
+                        if let Some(event_item) = event_item {
+                            metric!(
+                                counter(RelayCounters::ProcessingMessageProduced) += 1,
+                                event_type = &event_item.ty().to_string()
+                            );
+                        }
+                    }
+                }
             }
         }
 
         Ok(())
     }
 
+    fn extract_kafka_messages(
+        event_item: Option<&Item>,
+        event_id: EventId,
+        scoping: Scoping,
+        start_time: Instant,
+        remote_addr: Option<String>,
+        attachments: Vec<ChunkedAttachment>,
+    ) -> impl Iterator<Item = KafkaMessage> {
+        // There might be a better way to do this:
+        let (individual_attachments, inline_attachments) = match event_item {
+            Some(event_item) => {
+                if matches!(event_item.ty(), ItemType::Transaction) {
+                    (attachments, vec![])
+                } else {
+                    (vec![], attachments)
+                }
+            }
+            None => (attachments, vec![]),
+        };
+
+        let project_id = scoping.project_id;
+
+        let event_iterator = event_item
+            .map(|event_item| {
+                KafkaMessage::Event(EventKafkaMessage {
+                    payload: event_item.payload(),
+                    start_time: UnixTimestamp::from_instant(start_time).as_secs(),
+                    event_id,
+                    project_id,
+                    remote_addr,
+                    attachments: inline_attachments,
+                })
+            })
+            .into_iter();
+
```

Comment:
Let's restore these log lines:

```suggestion

if ! individual_attachments.empty() {
    relay_log::trace!("Sending individual attachments of envelope to kafka");
}

```